### PR TITLE
feat(sequencer): implement bridge deposits for incoming ICS20 transfers

### LINKED ISSUE #839: sequencer: ICS20 transfers to bridge accounts
allow ICS20 transfers to bridge accounts and emit an `Deposit`. require the packet memo to be the rollup destination address

### PR DESCRIPTION
**Pull request by noot**
## Summary
implement bridge deposits for incoming ICS20 transfers. this allows users to "directly" transfer an asset to a rollup via IBC. they must specify the rollup destination address in the ICS20 packet memo field.

## Background
users can transfer directly from another IBC-enabled chain to a rollup via an IBC transfer to the sequencer.

## Changes
- update `execute_ics20_transfer` to check if the transfer is to a bridge account; if so, ensure the packet memo field is set and the asset is valid
-  if the transfer is to a bridge account, emit a `Deposit` event during execution

## Testing
unit tests

## Related Issues

closes #839

### CODE REVIEW

File: crates/astria-sequencer/src/ibc/ics20_transfer.rs [hunk 1]
```diff
@@ -327,13 +330,63 @@ async fn execute_ics20_transfer<S: StateWriteExt>(
     // if the asset is prefixed with `ibc`, the rest of the denomination string is the asset ID,
     // so we need to look up the full trace from storage.
     // see https://github.com/cosmos/ibc-go/blob/main/docs/architecture/adr-001-coin-source-tracing.md#decision
-    if denom.prefix_is("ibc") {
+    if denom.prefix().starts_with("ibc") {
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/ibc/ics20_transfer.rs:333]: This is upstream code, right? This is much nicer.

File: crates/astria-sequencer/src/ibc/ics20_transfer.rs [hunk 2]
```diff
@@ -430,4 +474,160 @@ mod test {
         let asset = Denom::from("other_port/source_channel/asset".to_string());
         assert!(!is_prefixed(&source_port, &source_channel, &asset));
     }
+
+    #[tokio::test]
+    async fn execute_ics20_transfer_to_eoa() {
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/ibc/ics20_transfer.rs:479]: What does `to_eoa` mean? What does this test?

File: crates/astria-sequencer/src/ibc/ics20_transfer.rs [hunk 3]
```diff
@@ -327,13 +330,63 @@ async fn execute_ics20_transfer<S: StateWriteExt>(
...
+    //
+    // also, ensure that the asset ID being transferred
+    // to it is allowed.
+    let maybe_recipient_rollup_id = state
+        .get_bridge_account_rollup_id(&recipient)
+        .await
+        .context("failed to get bridge account rollup ID from state")?;
+    let is_bridge_lock = maybe_recipient_rollup_id.is_some();
+
+    // note: bridge accounts *are* allowed to do ICS20 withdrawals,
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/ibc/ics20_transfer.rs:352]: With the addition of this code, this method is growing very long and is hard to follow. Can you factor this out into self-standing functions?

File: crates/astria-sequencer/src/ibc/ics20_transfer.rs [hunk 4]
```diff
@@ -363,14 +416,10 @@ async fn execute_ics20_transfer<S: StateWriteExt>(
...
+
         state
-            .put_account_balance(
-                recipient,
-                denom.id(),
-                user_balance
-                    .checked_add(packet_amount)
-                    .ok_or(anyhow::anyhow!("overflow when adding to user balance"))?,
-            )
+            .increase_balance(recipient, denom.id(), packet_amount)
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/ibc/ics20_transfer.rs:421]: I like these methods, they signal the intent much better.

Can we prevent direct access to `state` and instead provide a wrapper that only exposes methods like `increase_balance`?

File: crates/astria-sequencer/src/ibc/ics20_transfer.rs [hunk 5]
```diff
@@ -430,4 +474,160 @@ mod test {
...
+
+        // use empty memo, which should fail
+        let packet = FungibleTokenPacketData {
+            denom: "nootasset".to_string(),
+            sender: String::new(),
+            amount: "100".to_string(),
+            receiver: hex::encode(bridge_address),
+            memo: String::new(),
+        };
+        let packet_bytes = serde_json::to_vec(&packet).expect("failed to serialize packet data");
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/ibc/ics20_transfer.rs:593]: This line states that you expec tthe serialization to fail. Is that right?

File: crates/astria-sequencer/src/ibc/ics20_transfer.rs [hunk 6]
```diff
@@ -430,4 +474,160 @@ mod test {
...
+                &mut state_tx,
+                &packet_bytes,
+                &"source_port".to_string().parse().unwrap(),
+                &"source_channel".to_string().parse().unwrap(),
+                &"dest_port".to_string().parse().unwrap(),
+                &"dest_channel".to_string().parse().unwrap(),
+                false,
+            )
+            .await
+            .is_err()
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/ibc/ics20_transfer.rs:630]: As above, consider using `expect_err` to document why this should have failed.

File: crates/astria-sequencer/src/ibc/ics20_transfer.rs [hunk 7]
```diff
@@ -430,4 +474,160 @@ mod test {
...
+
+        // use invalid asset, which should fail
+        let packet = FungibleTokenPacketData {
+            denom: "fake".to_string(),
+            sender: String::new(),
+            amount: "100".to_string(),
+            receiver: hex::encode(bridge_address),
+            memo: "destinationaddress".to_string(),
+        };
+        let packet_bytes = serde_json::to_vec(&packet).expect("failed to serialize packet data");
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/ibc/ics20_transfer.rs:617]: This `expect` message seems wrong.

File: crates/astria-sequencer/src/ibc/ics20_transfer.rs [hunk 8]
```diff
@@ -430,4 +474,160 @@ mod test {
...
+            .unwrap();
+
+        let packet = FungibleTokenPacketData {
+            denom: "nootasset".to_string(),
+            sender: String::new(),
+            amount: "100".to_string(),
+            receiver: hex::encode(bridge_address),
+            memo: "destinationaddress".to_string(),
+        };
+        let packet_bytes = serde_json::to_vec(&packet).expect("failed to serialize packet data");
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/ibc/ics20_transfer.rs:543]: wrong expect

File: crates/astria-sequencer/src/ibc/ics20_transfer.rs [hunk 9]
```diff
@@ -430,4 +474,160 @@ mod test {
...
+            &mut state_tx,
+            &packet_bytes,
+            &"source_port".to_string().parse().unwrap(),
+            &"source_channel".to_string().parse().unwrap(),
+            &"dest_port".to_string().parse().unwrap(),
+            &"dest_channel".to_string().parse().unwrap(),
+            false,
+        )
+        .await
+        .unwrap();
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/ibc/ics20_transfer.rs:555]:
```suggestion
        .expect("the ics20 transfer should have worked because ...");
```

File: crates/astria-sequencer/src/ibc/ics20_transfer.rs [hunk 10]
```diff
@@ -430,4 +474,160 @@ mod test {
...
+
+        let denom: Denom = format!("dest_port/dest_channel/{}", "nootasset").into();
+        let balance = state_tx
+            .get_account_balance(bridge_address, denom.id())
+            .await
+            .unwrap();
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/ibc/ics20_transfer.rs:561]:
```suggestion
            .expect("there should be a balance because the ics20 transfer moved funds of the given denomination to that address);
```
This would help a future maintainer understand what the intent of this check is.

File: crates/astria-sequencer/src/ibc/ics20_transfer.rs [hunk 11]
```diff
@@ -430,4 +474,160 @@ mod test {
...
+        assert_eq!(balance, 100);
+
+        let deposit = state_tx.get_block_deposits().await.unwrap();
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/ibc/ics20_transfer.rs:564]: as above

**Review by SuperFluffy — COMMENTED** [review]
The new logic looks correct, but it's hard to grok - also because the method it's embedded in is growing very long. Can you refactor it?

I really like the extra tests - but could please make more use of `.expect` (or `.expect_err`) and document why method calls should have passed/failed? That will make it easier for future maintainers understand what went wrong.

File: crates/astria-sequencer/src/ibc/ics20_transfer.rs (hunk 1 above)
Comment (noot) [reply crates/astria-sequencer/src/ibc/ics20_transfer.rs:333]: no this is our code! but yeah this is better/ actually correct

File: crates/astria-sequencer/src/ibc/ics20_transfer.rs (hunk 2 above)
Comment (noot) [reply crates/astria-sequencer/src/ibc/ics20_transfer.rs:479]: to externally owned account, so not a bridge account. this is testing an IBC transfer coming in to a non-bridge account

File: crates/astria-sequencer/src/ibc/ics20_transfer.rs (hunk 4 above)
Comment (noot) [reply crates/astria-sequencer/src/ibc/ics20_transfer.rs:421]: yeah, it's kind of out of scope for this PR but can do in a follow up

File: crates/astria-sequencer/src/ibc/ics20_transfer.rs (hunk 5 above)
Comment (noot) [reply crates/astria-sequencer/src/ibc/ics20_transfer.rs:593]: should be the opposite lol, will clarify

**Comment by noot** [comment]
@SuperFluffy addressed comments and refactored `execute_ics20_transfer` to have the bridge logic in a separate function!

File: crates/astria-sequencer/src/ibc/ics20_transfer.rs (hunk 1 above)
Comment (SuperFluffy) [reply crates/astria-sequencer/src/ibc/ics20_transfer.rs:400]: I see - yeah, that makes sense. Maybe we can remove the `prefix_is` method from core?

File: crates/astria-sequencer/src/ibc/ics20_transfer.rs (hunk 2 above)
Comment (SuperFluffy) [reply crates/astria-sequencer/src/ibc/ics20_transfer.rs:509]: Can you `s/eoa/externally_owned_account`? 🙏

File: crates/astria-sequencer/src/ibc/ics20_transfer.rs [hunk 12]
```diff
@@ -430,4 +504,159 @@ mod test {
...
+        let mut state_tx = StateDelta::new(snapshot.clone());
+
+        let packet = FungibleTokenPacketData {
+            denom: "nootasset".to_string(),
+            sender: String::new(),
+            amount: "100".to_string(),
+            receiver: "1c0c490f1b5528d8173c5de46d131160e4b2c0c3".to_string(),
+            memo: String::new(),
+        };
+        let packet_bytes = serde_json::to_vec(&packet).expect("failed to serialize packet data");
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/ibc/ics20_transfer.rs:523]: Note here that the intent is clear (compared to the more "test"-like unwraps below). So I'd be totally happy if you just `unwrap`ed here. It's pretty obvious that you just want serialize to json and that that's expected to work. :)

File: crates/astria-sequencer/src/ibc/ics20_transfer.rs [hunk 13]
```diff
@@ -430,4 +504,159 @@ mod test {
...
+            &"source_port".to_string().parse().unwrap(),
+            &"source_channel".to_string().parse().unwrap(),
+            &"dest_port".to_string().parse().unwrap(),
+            &"dest_channel".to_string().parse().unwrap(),
+            false,
+        )
+        .await
+        .unwrap();
+
+        let recipient = Address::try_from_slice(
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/ibc/ics20_transfer.rs:537]: Follow-up: should I add a `Address::from_hex` constructor?

File: crates/astria-sequencer/src/ibc/ics20_transfer.rs [hunk 14]
```diff
@@ -430,4 +504,159 @@ mod test {
...
+            &mut state_tx,
+            &packet_bytes,
+            &"source_port".to_string().parse().unwrap(),
+            &"source_channel".to_string().parse().unwrap(),
+            &"dest_port".to_string().parse().unwrap(),
+            &"dest_channel".to_string().parse().unwrap(),
+            false,
+        )
+        .await
+        .unwrap();
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/ibc/ics20_transfer.rs:535]: Here I'd add an `expect` message

File: crates/astria-sequencer/src/ibc/ics20_transfer.rs [hunk 15]
```diff
@@ -430,4 +504,159 @@ mod test {
...
+            .await
+            .unwrap();
+        assert_eq!(balance, 100);
+    }
+
+    #[tokio::test]
+    async fn execute_ics20_transfer_to_bridge_account_ok() {
+        let storage = cnidarium::TempStorage::new()
+            .await
+            .expect("failed to create temp storage backing chain state");
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/ibc/ics20_transfer.rs:553]: This message too is wrong

File: crates/astria-sequencer/src/ibc/ics20_transfer.rs [hunk 16]
```diff
@@ -430,4 +504,159 @@ mod test {
         let asset = Denom::from("other_port/source_channel/asset".to_string());
         assert!(!is_prefixed(&source_port, &source_channel, &asset));
     }
+
+    #[tokio::test]
+    async fn execute_ics20_transfer_to_eoa() {
+        let storage = cnidarium::TempStorage::new()
+            .await
+            .expect("failed to create temp storage backing chain state");
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/ibc/ics20_transfer.rs:512]: This message here is wrong as well. IMO the intent here is clear - you just want a temp storage. Feel free to `unwrap` here.

File: crates/astria-sequencer/src/ibc/ics20_transfer.rs [hunk 17]
```diff
@@ -430,4 +504,159 @@ mod test {
...
+
+        let recipient = Address::try_from_slice(
+            &hex::decode("1c0c490f1b5528d8173c5de46d131160e4b2c0c3").unwrap(),
+        )
+        .unwrap();
+        let denom: Denom = format!("dest_port/dest_channel/{}", "nootasset").into();
+        let balance = state_tx
+            .get_account_balance(recipient, denom.id())
+            .await
+            .unwrap();
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/ibc/ics20_transfer.rs:545]: Here I'd add an `expect` - why should this account have a balance? You are never explicitly `put`ing anything in here, so I have no way to know why it should be populated. Probably `.expect("the successful ics20 transfer should have added funds to this account")`?

File: crates/astria-sequencer/src/ibc/ics20_transfer.rs [hunk 18]
```diff
@@ -430,4 +504,159 @@ mod test {
...
+            .await
+            .expect("failed to create temp storage backing chain state");
+        let snapshot = storage.latest_snapshot();
+        let mut state_tx = StateDelta::new(snapshot.clone());
+
+        let bridge_address = Address::from([99; 20]);
+        let rollup_id = RollupId::from_unhashed_bytes(b"testchainid");
+        let denom: Denom = "nootasset".to_string().into();
+
+        state_tx.put_bridge_account_rollup_id(&bridge_address, &rollup_id);
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/ibc/ics20_transfer.rs:561]: This is infallible? Just caught my eye, the others are fallible.

File: crates/astria-sequencer/src/ibc/ics20_transfer.rs (hunk 4 above)
Comment (SuperFluffy) [reply crates/astria-sequencer/src/ibc/ics20_transfer.rs:451]: Yeah, that makes sense. I would love that

File: crates/astria-sequencer/src/ibc/ics20_transfer.rs [hunk 19]
```diff
@@ -430,4 +504,159 @@ mod test {
...
+        state_tx
+            .put_bridge_account_asset_ids(&bridge_address, &[denom.id()])
+            .unwrap();
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/ibc/ics20_transfer.rs:564]: Here `unwrap` is IMO totally fine - this happens during set up of the test, and looking at the function name you just want to dump some information into the `state`.

File: crates/astria-sequencer/src/ibc/ics20_transfer.rs [hunk 20]
```diff
@@ -430,4 +504,159 @@ mod test {
...
+
+        let packet = FungibleTokenPacketData {
+            denom: "nootasset".to_string(),
+            sender: String::new(),
+            amount: "100".to_string(),
+            receiver: hex::encode(bridge_address),
+            memo: "destinationaddress".to_string(),
+        };
+        let packet_bytes = serde_json::to_vec(&packet).expect("can serialize packet data");
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/ibc/ics20_transfer.rs:573]: Can `unwrap`, intent is clear

File: crates/astria-sequencer/src/ibc/ics20_transfer.rs [hunk 21]
```diff
@@ -430,4 +504,159 @@ mod test {
...
+            &mut state_tx,
+            &packet_bytes,
+            &"source_port".to_string().parse().unwrap(),
+            &"source_channel".to_string().parse().unwrap(),
+            &"dest_port".to_string().parse().unwrap(),
+            &"dest_channel".to_string().parse().unwrap(),
+            false,
+        )
+        .await
+        .expect("valid ics20 transfer to bridge account; recipient, memo, and asset ID are valid");
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/ibc/ics20_transfer.rs:585]: yeah, nice

File: crates/astria-sequencer/src/ibc/ics20_transfer.rs [hunk 22]
```diff
@@ -430,4 +504,159 @@ mod test {
...
+
+        let denom: Denom = format!("dest_port/dest_channel/{}", "nootasset").into();
+        let balance = state_tx
+            .get_account_balance(bridge_address, denom.id())
+            .await
+            .expect("ics20 transfer from sender to bridge account should succeed");
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/ibc/ics20_transfer.rs:591]: Maybe?
```suggestion
            .expect("ics20 transfer from sender to bridge account should have updated funds in the bridge address");
```

File: crates/astria-sequencer/src/ibc/ics20_transfer.rs [hunk 23]
```diff
@@ -430,4 +504,159 @@ mod test {
...
+            .await
+            .expect("a deposit should exist as a result of the transfer to a bridge account");
+        assert_eq!(deposit.len(), 1);
+    }
+
+    #[tokio::test]
+    async fn execute_ics20_transfer_to_bridge_account_invalid() {
+        let storage = cnidarium::TempStorage::new()
+            .await
+            .expect("failed to create temp storage backing chain state");
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/ibc/ics20_transfer.rs:605]: `unwrap` would be ok

File: crates/astria-sequencer/src/ibc/ics20_transfer.rs [hunk 24]
```diff
@@ -430,4 +504,159 @@ mod test {
...
+        let mut state_tx = StateDelta::new(snapshot.clone());
+
+        let bridge_address = Address::from([99; 20]);
+        let rollup_id = RollupId::from_unhashed_bytes(b"testchainid");
+        let denom: Denom = "nootasset".to_string().into();
+
+        state_tx.put_bridge_account_rollup_id(&bridge_address, &rollup_id);
+        state_tx
+            .put_bridge_account_asset_ids(&bridge_address, &[denom.id()])
+            .unwrap();
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/ibc/ics20_transfer.rs:616]: that's also ok IMO

File: crates/astria-sequencer/src/ibc/ics20_transfer.rs [hunk 25]
```diff
@@ -430,4 +504,159 @@ mod test {
...
+
+        // use empty memo, which should fail
+        let packet = FungibleTokenPacketData {
+            denom: "nootasset".to_string(),
+            sender: String::new(),
+            amount: "100".to_string(),
+            receiver: hex::encode(bridge_address),
+            memo: String::new(),
+        };
+        let packet_bytes = serde_json::to_vec(&packet).expect("can serialize packet data");
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/ibc/ics20_transfer.rs:626]: can unwrap

File: crates/astria-sequencer/src/ibc/ics20_transfer.rs [hunk 26]
```diff
@@ -430,4 +504,159 @@ mod test {
...
+            &mut state_tx,
+            &packet_bytes,
+            &"source_port".to_string().parse().unwrap(),
+            &"source_channel".to_string().parse().unwrap(),
+            &"dest_port".to_string().parse().unwrap(),
+            &"dest_channel".to_string().parse().unwrap(),
+            false,
+        )
+        .await
+        .expect_err("empty memo field during transfer to bridge account should fail");
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/ibc/ics20_transfer.rs:638]: Nice, maybe also:
```suggestion
        .expect_err("empty packet memo field during transfer to bridge account should fail");
```

File: crates/astria-sequencer/src/ibc/ics20_transfer.rs [hunk 27]
```diff
@@ -430,4 +504,159 @@ mod test {
...
+
+        // use invalid asset, which should fail
+        let packet = FungibleTokenPacketData {
+            denom: "fake".to_string(),
+            sender: String::new(),
+            amount: "100".to_string(),
+            receiver: hex::encode(bridge_address),
+            memo: "destinationaddress".to_string(),
+        };
+        let packet_bytes = serde_json::to_vec(&packet).expect("can serialize packet data");
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/ibc/ics20_transfer.rs:648]: unwrap ok

File: crates/astria-sequencer/src/ibc/ics20_transfer.rs [hunk 28]
```diff
@@ -430,4 +504,159 @@ mod test {
...
+
+        let deposit = state_tx
+            .get_block_deposits()
+            .await
+            .expect("a deposit should exist as a result of the transfer to a bridge account");
+        assert_eq!(deposit.len(), 1);
+    }
+
+    #[tokio::test]
+    async fn execute_ics20_transfer_to_bridge_account_invalid() {
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/ibc/ics20_transfer.rs:602]: This thing contains multiple tests - could you break them out into individual tests? By the time I reached the end I was no longer sure which test I am in.

File: crates/astria-sequencer/src/state_ext.rs [hunk 29]
```diff
@@ -243,3 +243,26 @@ pub(crate) trait StateWriteExt: StateWrite {
...
+mod test {
+    use cnidarium::StateDelta;
+
+    use super::{
+        StateReadExt as _,
+        StateWriteExt as _,
+    };
+
+    #[tokio::test]
+    async fn chain_id() {
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/state_ext.rs:257]:
```suggestion
    async fn chain_id_to_storage_roundtrip() {
```
or
```suggestion
    async fn chain_id_storage_write_read() {
```

File: crates/astria-sequencer/src/ibc/ics20_transfer.rs [hunk 30]
```diff
@@ -303,6 +306,73 @@ impl AppHandlerExecute for Ics20Transfer {
...
+    // check if the recipient is a bridge account; if so,
+    // ensure that the packet memo field (`destination_address`) is set.
+    //
+    // also, ensure that the asset ID being transferred
+    // to it is allowed.
+    let maybe_recipient_rollup_id = state
+        .get_bridge_account_rollup_id(recipient)
+        .await
+        .context("failed to get bridge account rollup ID from state")?;
+    let is_bridge_lock = maybe_recipient_rollup_id.is_some();
```
Comment (SuperFluffy) [inline crates/astria-sequencer/src/ibc/ics20_transfer.rs:326]: This and the checks below could just be rolled into one test:
```rust
    if maybe_recipient_rollup_id.is_none() || is_refund {
        return Ok(());
    }
```

File: crates/astria-sequencer/src/ibc/ics20_transfer.rs (hunk 13 above)
Comment (noot) [reply crates/astria-sequencer/src/ibc/ics20_transfer.rs:537]: yeah that would be useful

File: crates/astria-sequencer/src/state_ext.rs (hunk 29 above)
Comment (noot) [reply crates/astria-sequencer/src/state_ext.rs:257]: omg sorry i committed this by accident

**Comment by noot** [comment]
@SuperFluffy comments addressed!

**Review by SuperFluffy — APPROVED** [review]
This PR is great, thank you for addressing the comments.

The only bit I'd ask for is not abbreviating `eoa` because that terminology is not at all obvious. :)
